firstvoices/backend/serializers/song_serializers.py

```diff
@@ -16,18 +21,50 @@ class LyricSerializer(ModelSerializer):
     class Meta:
         model = Lyric
         fields = ("id", "text", "translation")
+        read_only_fields = ("id",)
 
 
-class SongDetailSerializer(
-    RelatedMediaSerializerMixin, SiteContentLinkedTitleSerializer
+class SongSerializer(
+    UpdateSerializerMixin,
+    CreateSiteContentSerializerMixin,
+    RelatedMediaSerializerMixin,
+    SiteContentLinkedTitleSerializer,
 ):
-    cover_image = ImageSerializer()
-    site = LinkedSiteSerializer()
-
+    cover_image = WriteableRelatedImageSerializer(
+        allow_null=True, queryset=Image.objects.all()
+    )
+    site = LinkedSiteSerializer(required=False, read_only=True)
     lyrics = LyricSerializer(many=True)
 
+    def create(self, validated_data):
+        lyrics = validated_data.pop("lyrics")
+
+        created = super().create(validated_data)
+
+        for index, lyric_data in enumerate(lyrics):
+            Lyric.objects.create(song=created, ordering=index, **lyric_data)
+
+        return created
+
+    def update(self, instance, validated_data):
+        Lyric.objects.filter(song__id=instance.id).delete()
+        try:
+            lyrics = validated_data.pop("lyrics")
+            for index, lyric_data in enumerate(lyrics):
+                Lyric.objects.create(song=instance, ordering=index, **lyric_data)
+        except KeyError:
+            pass
+
+        return super().update(instance, validated_data)
+
     class Meta(SiteContentLinkedTitleSerializer.Meta):
         model = Song
+        read_only_fields = (
+            base_id_fields,
+            base_timestamp_fields,
+            "cover_image",
```

Comment:
Why is the cover_image read-only?

Reply:
that was a remnant of a previous version. it had no effect but I've removed it for clarity.